### PR TITLE
spider fixes

### LINKED ISSUE #946: spider protocols fix
1. ap sr must support threading
2. change csh to sh, since there are csh versions that do not work with currect script synthax

### PR DESCRIPTION
**Pull request by azazellochg**
1. threading now works in most protocols (refinement not tested since there is no test protocol)
2. change csh to sh for more compatibility
3. add output generation in spider reconstruct protocol
Closes #946

### CODE REVIEW

File: pyworkflow/em/packages/spider/scripts/mda/apsr4class.msa [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ; Using SPIDER command AP SR
 
 ; ---------------- Parameters ----------------
+x30 = 0                ; number of threads to use, 0 = use all
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/spider/scripts/mda/apsr4class.msa]: @azazellochg Do you mind using [nummps] here? Just to provide a more clear name and be consistent with the other scripts

Comment (azazellochg) [inline pyworkflow/em/packages/spider/scripts/mda/apsr4class.msa]: @delarosatrevin sure, you can modify that.

**Comment by delarosatrevin** [comment]
I have passed all test related to Spider and seems to be working, except:
1) The test: scipion test tests.em.workflows.test_workflow_spiderMDA.TestSpiderConvert is failing, but this not seems to be related with these changes
2) Even if the MPI=0 in the addParallelSection, it is been display in the form...but this seems like a general issue.

@pconesa Can you check about 2) ? or maybe @ldelcano that changed to 1 or 0 recently.

**Comment by azazellochg** [comment]
@delarosatrevin hm, for me this test works..

**Comment by pconesa** [comment]
Test works for me too. I'll check the MPI thing now.

**Comment by pconesa** [comment]
If I'm right, @ldelcano only reviewed all protocols that has STEPS_PARALLEL and none of the Spider protocols are parallelized this way. We haven't touched the form, so it might be an older issue. What should be the behavior in the form for protocols not parallelized by scipion?:
None in any (Thread or MPI should hide the field?)

**Comment by delarosatrevin** [comment]
It has been a long time since I have modified the form.py core file, but I think if any of threads or mpi is equal to 0 in the definition of the parameters (addParallelSection), they should not appear in the form. Zero means that are not allowed.

**Comment by pconesa** [comment]
As long as the protocol has the attribute it will show up:

`self.allowMpi = hasattr(self, 'numberOfMpi')`

And in addParallelSection only if mpi > 0 it will add the parameter. So you are guessing rith @delarosatrevin . I guess we want no MPI field in these spider protocols, right?

**Comment by pconesa** [comment]
Issue is here...and probably with all protocols that have multiple inheritance, or at least invoking 2 __init__:

`class SpiderProtAlign(ProtAlign2D, SpiderProtocol):
    """ Base class for Spider alignment protocols. """
    
    def __init__(self, script, alignDir, **args):
        ProtAlign2D.__init__(self, **args)
        SpiderProtocol.__init__(self, **args)
`

With the first __init__ call , allowMpi is set correctly to false, but at the same time numberOfMpi is set to 0.
Within the second __init__ the object has the attribute allready numberOfMpi set to 0 and therefore allowMpi is set to True.

**Comment by pconesa** [comment]
@delarosatrevin this will fix the issue for multiple inheritance protocols:

`       # NOTE: For multiple inheritance protocols code is passing here twice
        # and second time numberOfMpi attribute is set to 1.
        if not hasattr(self, 'allowMpi'):
            self.allowMpi = hasattr(self, 'numberOfMpi')`

I feel is a bit of a hack....but behaves and seems safe enough. Any cleaner fix?

Same for threads.

**Comment by delarosatrevin** [comment]
I'm not sure...what about if the second pass is the proper one? This will make us to review now all multiple inheritance. Anyway I think the way it is now is not good.

**Comment by pconesa** [comment]
Yep, you are right, I realized that while driving back.

We'll review how many multiple inheritance there are and will see.

May be for this we can set the allowMpi to false after the second init 
and review it later with the rest.

What do you think?

**Comment by delarosatrevin** [comment]
Yep, I will here focus in Spider protocols (for most of which I think they do not support MPI) and then think in a more general solution for the future.

**Comment by azazellochg** [comment]
@delarosatrevin indeed, very few support mpi natively: https://spider.wadsworth.org/spider_doc/spider/docs/mpi.html

**Comment by pconesa** [comment]
Thanks @azazellochg and @delarosatrevin , looking at your the link.....I see that we need to use "spider_linux_mpi_opt64" and this is not found in our code. Can I assume that none of our Spider protocols are currently implemented to make use of the MPI capabilities of Spider'

**Comment by azazellochg** [comment]
@pconesa scpider-reconstruct and spider-refinement could use mpi, but it's not implemented for now.

**Comment by pconesa** [comment]
Thanks Gregory! I think MPI functionality will not go in the 
release....unless any of you insist and see it very, very clear and safe.

How do you see it? Is it obvious?

**Comment by delarosatrevin** [comment]
I would say that it is not important for now.

**Comment by azazellochg** [comment]
i would leave them "as is" for now, since they are not even fully finished (using dummy zero angles etc)

**Comment by pconesa** [comment]
So, all of the Spider protocols show now only "Threads".

**Pull request merged**